### PR TITLE
Flexible id tags

### PR DESCRIPTION
**Pull request by sukhwinder33445**
## Blocked by

- https://github.com/Icinga/icinga-notifications/pull/105
- https://github.com/Icinga/ipl-orm/pull/122

### CODE REVIEW

**Comment by yhabteab** [comment]
Having this custom tags without the `service` tag being defined:
Bildschirmfoto 2023-10-09 um 12 36 41

Web renders these entries everywhere as follows:
Bildschirmfoto 2023-10-09 um 12 39 39

**Note:** It's a host object not service

**Comment by yhabteab** [comment]
Bildschirmfoto 2023-10-09 um 14 21 48

File: library/Notifications/Widget/Detail/EventDetail.php [hunk 1]
```diff
@@ -109,6 +103,10 @@ protected function createIncident(): ?array
             return null;
         }
 
+        /** @var Query $ObjectIdTagQuery */
+        $ObjectIdTagQuery = $this->event->incident->object->object_id_tag;
+        $ObjectIdTagQuery->getSelectBase()->resetWhere();
+        $ObjectIdTagQuery->filter(Filter::equal('object_id', $this->event->object_id));
```
Comment (nilmerg) [inline library/Notifications/Widget/Detail/EventDetail.php:109]: Unused, remove this

File: library/Notifications/Widget/ItemList/IncidentHistoryListItem.php [hunk 2]
```diff
@@ -53,23 +55,16 @@ protected function assembleTitle(BaseHtmlElement $title): void
...
-                $content = new Link(
-                    $event->object->host,
-                    Links::event($this->item->event_id),
-                    ['class' => 'subject']
-                );
-            }
+            /** @var Query $ObjectIdTagQuery */
+            $ObjectIdTagQuery = $event->object->object_id_tag;
+            $ObjectIdTagQuery->getSelectBase()->resetWhere();
+            $ObjectIdTagQuery->filter(Filter::equal('object_id', $event->object_id));
```
Comment (nilmerg) [inline library/Notifications/Widget/ItemList/IncidentHistoryListItem.php:61]: Unused, remove this

File: library/Notifications/Model/Objects.php [hunk 3]
```diff
@@ -60,7 +50,20 @@ public function createRelations(Relations $relations)
...
         $relations->hasMany('incident', Incident::class);
         $relations->belongsTo('source', Source::class);
+        $relations->hasMany('object_id_tag', ObjectIdTag::class);
         $relations->hasMany('object_extra_tag', ObjectExtraTag::class)
             ->setJoinType('LEFT');
     }
+
+    //TODO: Just a temporary solution to avoid changing code everywhere.
+    // This must be done with hooks, remove once hooks are available
+    public function getTagsList(): string
```
Comment (nilmerg) [inline library/Notifications/Model/Objects.php:60]: Rename to `getName()` please.

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: library/Notifications/Widget/Detail/EventDetail.php (hunk 1 above)
Comment (nilmerg) [reply library/Notifications/Widget/Detail/EventDetail.php:109]: With https://github.com/Icinga/ipl-orm/pull/122, this is not required anymore

File: library/Notifications/Widget/ItemList/IncidentHistoryListItem.php (hunk 2 above)
Comment (nilmerg) [reply library/Notifications/Widget/ItemList/IncidentHistoryListItem.php:61]: With https://github.com/Icinga/ipl-orm/pull/122, this is not required anymore

**Comment by julianbrost** [comment]
I wanted to give this a try, but I only got exceptions on the event and incident detail pages.

Event detail:
```
SQLSTATE[42P01]: Undefined table: 7 ERROR: missing FROM-clause entry for table "event"
LINE 1: ...d_tag.object_id WHERE sub_object_id_tag_object.id = event.id
^

#0 /usr/share/icinga-php/ipl/vendor/ipl/sql/src/Connection.php(402): PDOStatement->execute()
#1 /usr/share/icinga-php/ipl/vendor/ipl/sql/src/Connection.php(418): ipl\Sql\Connection->prepexec()
#2 /usr/share/icinga-php/ipl/vendor/ipl/orm/src/Query.php(699): ipl\Sql\Connection->select()
#3 /usr/share/icinga-php/ipl/vendor/ipl/orm/src/ResultSet.php(142): ipl\Orm\Query->yieldResults()
#4 [internal function]: ipl\Orm\ResultSet->yieldTraversable()
#5 /usr/share/icinga-php/ipl/vendor/ipl/orm/src/ResultSet.php(122): Generator->valid()
#6 /usr/share/icinga-php/ipl/vendor/ipl/orm/src/ResultSet.php(114): ipl\Orm\ResultSet->advance()
#7 /usr/share/icingaweb2/modules/notifications/library/Notifications/Model/Objects.php(69): ipl\Orm\ResultSet->rewind()
#8 /usr/share/icingaweb2/modules/notifications/library/Notifications/Widget/ItemList/IncidentListItem.php(51): Icinga\Module\Notifications\Model\Objects->getName()
#9 /usr/share/icingaweb2/modules/notifications/library/Notifications/Common/BaseListItem.php(111): Icinga\Module\Notifications\Widget\ItemList\IncidentListItem->assembleTitle()
#10 /usr/share/icingaweb2/modules/notifications/library/Notifications/Widget/ItemList/IncidentListItem.php(71): Icinga\Module\Notifications\Common\BaseListItem->createTitle()
#11 /usr/share/icingaweb2/modules/notifications/library/Notifications/Common/BaseListItem.php(84): Icinga\Module\Notifications\Widget\ItemList\IncidentListItem->assembleHeader()
#12 /usr/share/icingaweb2/modules/notifications/library/Notifications/Widget/ItemList/IncidentListItem.php(103): Icinga\Module\Notifications\Common\BaseListItem->createHeader()
#13 /usr/share/icingaweb2/modules/notifications/library/Notifications/Common/BaseListItem.php(93): Icinga\Module\Notifications\Widget\ItemList\IncidentListItem->assembleMain()
#14 /usr/share/icingaweb2/modules/notifications/library/Notifications/Common/BaseListItem.php(129): Icinga\Module\Notifications\Common\BaseListItem->createMain()
#15 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(344): Icinga\Module\Notifications\Common\BaseListItem->assemble()
#16 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(566): ipl\Html\HtmlDocument->ensureAssembled()
#17 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(390): ipl\Html\HtmlDocument->render()
#18 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(297): ipl\Html\HtmlDocument->renderUnwrapped()
#19 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(365): ipl\Html\BaseHtmlElement->renderContent()
#20 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(568): ipl\Html\BaseHtmlElement->renderUnwrapped()
#21 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(390): ipl\Html\HtmlDocument->render()
#22 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(297): ipl\Html\HtmlDocument->renderUnwrapped()
#23 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(365): ipl\Html\BaseHtmlElement->renderContent()
#24 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(568): ipl\Html\BaseHtmlElement->renderUnwrapped()
#25 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(390): ipl\Html\HtmlDocument->render()
#26 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(297): ipl\Html\HtmlDocument->renderUnwrapped()
#27 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(365): ipl\Html\BaseHtmlElement->renderContent()
#28 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(568): ipl\Html\BaseHtmlElement->renderUnwrapped()
#29 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(390): ipl\Html\HtmlDocument->render()
#30 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(568): ipl\Html\HtmlDocument->renderUnwrapped()
#31 /usr/share/icinga-php/ipl/vendor/ipl/web/src/Compat/ViewRenderer.php(56): ipl\Html\HtmlDocument->render()
#32 /usr/share/icinga-php/vendor/vendor/shardj/zf1-future/library/Zend/Controller/Action/Helper/ViewRenderer.php(970): ipl\Web\Compat\ViewRenderer->render()
#33 /usr/share/icinga-php/vendor/vendor/shardj/zf1-future/library/Zend/Controller/Action/HelperBroker.php(277): Zend_Controller_Action_Helper_ViewRenderer->postDispatch()
#34 /usr/share/icinga-php/vendor/vendor/shardj/zf1-future/library/Zend/Controller/Action.php(527): Zend_Controller_Action_HelperBroker->notifyPostDispatch()
#35 /usr/share/icingaweb2/library/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch()
#36 /usr/share/icinga-php/vendor/vendor/shardj/zf1-future/library/Zend/Controller/Front.php(954): Icinga\Web\Controller\Dispatcher->dispatch()
#37 /usr/share/icingaweb2/library/Icinga/Application/Web.php(294): Zend_Controller_Front->dispatch()
#38 /usr/share/icingaweb2/library/Icinga/Application/webrouter.php(105): Icinga\Application\Web->dispatch()
#39 /usr/share/icingaweb2/public/index.php(4): require_once(String)
#40 {main}
```

Incident detail:
```
SQLSTATE[42P01]: Undefined table: 7 ERROR: missing FROM-clause entry for table "sub_incident_history"
LINE 1: ...tag.object_id WHERE sub_object_id_tag_object.id = sub_incide...
^

#0 /usr/share/icinga-php/ipl/vendor/ipl/sql/src/Connection.php(402): PDOStatement->execute()
#1 /usr/share/icinga-php/ipl/vendor/ipl/sql/src/Connection.php(418): ipl\Sql\Connection->prepexec()
#2 /usr/share/icinga-php/ipl/vendor/ipl/orm/src/Query.php(699): ipl\Sql\Connection->select()
#3 /usr/share/icinga-php/ipl/vendor/ipl/orm/src/ResultSet.php(142): ipl\Orm\Query->yieldResults()
#4 [internal function]: ipl\Orm\ResultSet->yieldTraversable()
#5 /usr/share/icinga-php/ipl/vendor/ipl/orm/src/ResultSet.php(122): Generator->valid()
#6 /usr/share/icinga-php/ipl/vendor/ipl/orm/src/ResultSet.php(114): ipl\Orm\ResultSet->advance()
#7 /usr/share/icingaweb2/modules/notifications/library/Notifications/Model/Objects.php(69): ipl\Orm\ResultSet->rewind()
#8 /usr/share/icingaweb2/modules/notifications/library/Notifications/Widget/ItemList/IncidentHistoryListItem.php(59): Icinga\Module\Notifications\Model\Objects->getName()
#9 /usr/share/icingaweb2/modules/notifications/library/Notifications/Common/BaseListItem.php(111): Icinga\Module\Notifications\Widget\ItemList\IncidentHistoryListItem->assembleTitle()
#10 /usr/share/icingaweb2/modules/notifications/library/Notifications/Widget/ItemList/IncidentHistoryListItem.php(75): Icinga\Module\Notifications\Common\BaseListItem->createTitle()
#11 /usr/share/icingaweb2/modules/notifications/library/Notifications/Common/BaseListItem.php(84): Icinga\Module\Notifications\Widget\ItemList\IncidentHistoryListItem->assembleHeader()
#12 /usr/share/icingaweb2/modules/notifications/library/Notifications/Widget/ItemList/IncidentHistoryListItem.php(145): Icinga\Module\Notifications\Common\BaseListItem->createHeader()
#13 /usr/share/icingaweb2/modules/notifications/library/Notifications/Common/BaseListItem.php(93): Icinga\Module\Notifications\Widget\ItemList\IncidentHistoryListItem->assembleMain()
#14 /usr/share/icingaweb2/modules/notifications/library/Notifications/Common/BaseListItem.php(129): Icinga\Module\Notifications\Common\BaseListItem->createMain()
#15 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(344): Icinga\Module\Notifications\Common\BaseListItem->assemble()
#16 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(566): ipl\Html\HtmlDocument->ensureAssembled()
#17 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(390): ipl\Html\HtmlDocument->render()
#18 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(297): ipl\Html\HtmlDocument->renderUnwrapped()
#19 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(365): ipl\Html\BaseHtmlElement->renderContent()
#20 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(568): ipl\Html\BaseHtmlElement->renderUnwrapped()
#21 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(390): ipl\Html\HtmlDocument->render()
#22 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(297): ipl\Html\HtmlDocument->renderUnwrapped()
#23 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(365): ipl\Html\BaseHtmlElement->renderContent()
#24 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(568): ipl\Html\BaseHtmlElement->renderUnwrapped()
#25 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(390): ipl\Html\HtmlDocument->render()
#26 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(297): ipl\Html\HtmlDocument->renderUnwrapped()
#27 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(365): ipl\Html\BaseHtmlElement->renderContent()
#28 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(568): ipl\Html\BaseHtmlElement->renderUnwrapped()
#29 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(390): ipl\Html\HtmlDocument->render()
#30 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(568): ipl\Html\HtmlDocument->renderUnwrapped()
#31 /usr/share/icinga-php/ipl/vendor/ipl/web/src/Compat/ViewRenderer.php(56): ipl\Html\HtmlDocument->render()
#32 /usr/share/icinga-php/vendor/vendor/shardj/zf1-future/library/Zend/Controller/Action/Helper/ViewRenderer.php(970): ipl\Web\Compat\ViewRenderer->render()
#33 /usr/share/icinga-php/vendor/vendor/shardj/zf1-future/library/Zend/Controller/Action/HelperBroker.php(277): Zend_Controller_Action_Helper_ViewRenderer->postDispatch()
#34 /usr/share/icinga-php/vendor/vendor/shardj/zf1-future/library/Zend/Controller/Action.php(527): Zend_Controller_Action_HelperBroker->notifyPostDispatch()
#35 /usr/share/icingaweb2/library/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch()
#36 /usr/share/icinga-php/vendor/vendor/shardj/zf1-future/library/Zend/Controller/Front.php(954): Icinga\Web\Controller\Dispatcher->dispatch()
#37 /usr/share/icingaweb2/library/Icinga/Application/Web.php(294): Zend_Controller_Front->dispatch()
#38 /usr/share/icingaweb2/library/Icinga/Application/webrouter.php(105): Icinga\Application\Web->dispatch()
#39 /usr/share/icingaweb2/public/index.php(4): require_once(String)
#40 {main}
```

Tried with this PR as is as well as rebased onto main in both the `icinga/icingaweb2:2.12.0` and `icinga/icingaweb2:main` (however in an older version, as the current image doesn't even start) and all got those exceptions.

For the parts that I could try, the looks are more, let's say functional:

Event List
Incident List

That should probably receive some nicer styling, however, I don't decide if this is part of the scope of this PR.

**Comment by sukhwinder33445** [comment]
Please pull `icinga-php-library`'s `snapshot/nightly`.

**Review by nilmerg — CHANGES_REQUESTED** [review]
Adding new patterns to the phpstan baseline is already not a thing I'd accept, but some of them seem to relate to actual bugs. Please have a look and avoid adding stuff to the baseline.

**Review by nilmerg — APPROVED** [review]
I approve this, because none of the issues is a deal breaker. To get this done before my vacation, I accept these limitations. Though, they should be tackled later on.

**Pull request merged**